src/utilities/include/utils.h

```diff
@@ -213,5 +215,62 @@ class ConcurrentSharedPtr {
   std::shared_ptr<std::recursive_mutex> mutex =
       std::make_shared<std::recursive_mutex>();
 };
+
+template <typename I = int, typename R = unsigned char>
+struct UnionFind {
+  Vec<I> parents;
+  // we do union by rank
+  // note that we shift rank by 1, rank 0 means it is not connected to anything
+  // else
+  Vec<R> ranks;
+
+  UnionFind(I numNodes) : parents(numNodes), ranks(numNodes, 0) {
+    sequence(autoPolicy(numNodes), parents.begin(), parents.end());
+  }
+
+  I find(I x) {
+    while (parents[x] != x) {
+      parents[x] = parents[parents[x]];
+      x = parents[x];
+    }
+    return x;
+  }
+
+  void unionXY(I x, I y) {
+    if (x == y) return;
+    if (ranks[x] == 0) ranks[x] = 1;
+    if (ranks[y] == 0) ranks[y] = 1;
+    x = find(x);
+    y = find(y);
+    if (x == y) return;
+    if (ranks[x] < ranks[y]) std::swap(x, y);
+    if (ranks[x] == ranks[y]) ranks[x]++;
+    parents[y] = x;
+  }
+
+  I connectedComponents(std::vector<I>& components) {
+    components.resize(parents.size());
+    I lonelyNodes = 0;
+    std::unordered_map<I, I> toLabel;
```

Comment:
I can't remember if this is true everywhere, but I recall being annoyed that connected components returned arbitrary sequential labels instead of a representative member, so I ended up having to add code to get back to a representative member. It may be possible that the connected components part of this isn't actually necessary and that `find` may be all we really need.

Reply:
I didn't spend much time looking at the code so I only changed the algorithm for `MeshGL::Merge` that I can immediately recognize.

It will be much faster if you can just work with a representative member instead of requiring a sequential component ID (and don't need the component count). This `connectedComponent` is a lot slower than the union part. In fact, it probably take around 80% of the time in `GetLabel`.